### PR TITLE
hide more file writing details from engine

### PR DESCRIPTION
**Pull request by tabokie**
This is a preparing work for #258.

Changes:
- Move `fsync` to file handle instead of writer, so that we can call it concurrently in the future. In doing so, we have to remove the sync offset tracking and deprecate `bytes_per_sync` support
- Move `prepare_write` into pipe_log. Introduce a trait `ReactiveBytes` for this purpose.
- Rotate the file during write instead of doing it every write group, the timing is also delayed to the next write after exceeding limit (instead of the write that exceeds the limit).
- Differentiate two types of I/O errors: unrecoverable error from fsync, and retriable error from pwrite. Only bubble error for the latter case, panic early for unrecoverable ones.
- Also refactor the purge code, fix two cases where force-compact is not triggered

### CODE REVIEW

File: src/file_pipe_log/pipe.rs [hunk 1]
```diff
@@ -401,15 +413,11 @@ impl<F: FileSystem> SinglePipe<F> {
         Ok(handle)
     }
 
-    fn maybe_sync(&self, force: bool) -> Result<()> {
+    fn sync(&self) -> Result<()> {
```
Comment (LykxSassinator) [inline src/file_pipe_log/pipe.rs:416]: In current implementation of `sync`, it still gets the `mutex` for avoiding concurrent racing at the beginning.

According to #258 , will it be refactored to support concurrent accessing in later work ? i.e. replace `mutex` with `Rwlock` ? or add an extra `F::Handle` in `SinglePipe`, making it be called concurrently ?

**Review by LykxSassinator — COMMENTED** [review]
Rest is LGTM

File: src/file_pipe_log/pipe.rs (hunk 1 above)
Comment (tabokie) [reply src/file_pipe_log/pipe.rs:416]: Yes.

**Comment by tabokie** [comment]
Ping @Connor1996 , this one is blocking #261

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**